refactor(compiler): Generate trusted const values for extracted attrs

mmalerba opened the pull request.

mmalerba committed:
refactor(compiler): Generate trusted const values for extracted attrs

Use the DomElementSchemaRegistry to determine the correct security
context for static attributes, and pass it along during ingestion. Then
during the resolve sanitizers phase, use the security context to
determine if a trusted value function is needed

dylhunn commented on packages/compiler/src/template/pipeline/src/phases/parse_extracted_styles.ts:
```diff
@@ -6,9 +6,12 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {SecurityContext} from '@angular/compiler/src/core';
```
i think you need a relative import here

mmalerba replied:
ah yeah, auto-import got me

dylhunn commented on packages/compiler/src/template/pipeline/src/phases/reify.ts:
```diff
@@ -36,6 +41,16 @@ export function reify(job: CompilationJob): void {
+  }
+}
+
+function reifyConsts(job: ComponentCompilationJob) {
```
I'm not sure how I feel about extending the `reify` phase to perform in-place transformations on expressions in the const array. IMO, stuff in the const array should already be "finalized," and this design breaks that abstraction. Is there some way we could do this transformation in the const_collection phase, or in a small separate phase before they get const collected?

mmalerba replied:
Why do we need to consider it finalized? It seemed like a bug to me that we didn't already do this. All of our ops can refer to internal expressions up until the reify phase, so it made sense in my head that consts expressions should be  able to do the same.

As for why I did it this way, I think back when I first did sanitizers for style bindings reify was the only phase that imported `render3/r3_identifiers` and I wasn't sure if it was ok to use that in other phases, so I made an intermediate enum as a placeholder until we got to the reify phase. Now I see other phases using the identifiers as well, so if its ok to do that I'll happily clean up the intermediate enums and just push the final expression

dylhunn replied:
Hm, I don't know that I 100% agree but TBH I don't feel that strongly. If you think this is the clearest way, let's just go ahead with this.

mmalerba replied:
I changed it to just push the `importExpr` right away. That was just something I misunderstood back when doing the style binding sanitization. With that gone there's no real reason to transform the consts, so I removed it as an optimization. However, I still don't really understand why philosophically its ok to have ir expressions in our ops but not in our consts

dylhunn replied:
It looks good! I do prefer this, but actually every time I come back to this, I feel less and less strongly that consts need to be finalized at the time they are added. I just was used to thinking about the consts array as being an "output" structure, not an IR structure, but I think you're probably right that it would have been fine.

mmalerba committed:
refactor(compiler): Simplify how sanitizers are generated

Previously we generated an intermediate expression which was later
converted into a symbol import expression for the sanitizer function.
This commit simplifies the behavior by just generating the symbol import
from the beginning

dylhunn commented:
Caretaker: no impact on g3 presubmit

alxhub commented:
This PR was merged into the repository by commit 4f7d8f3bac36d540f6565ebec9b9049b041733a8.
